NomicFoundation/hardhat
# Add Hardhat Ignition to hardhat-toolbox

Updates the `@nomicfoundation/hardhat-toolbox` and `@nomicfoundation/hardhat-toolbox-viem` packages to include Hardhat Ignition.

This includes

- Updated sample projects with Hardhat Ignition files
- Added Hardhat Ignition to `hardhat-toolbox` and `hardhat-toolbox-viem`
- Updated relevant docs to use Hardhat Ignition where relevant


A sibling PR to update [hardhat-boilerplate](https://github.com/NomicFoundation/hardhat-boilerplate/pull/102) has also been opened.

resolves https://github.com/NomicFoundation/hardhat-ignition/issues/614
resolves https://github.com/NomicFoundation/hardhat-ignition/issues/613

## Manual Testing Instructions

To manually create HH projects with the new toolbox:

1. Pull the branch locally
2. Install the new deps: `pnpm install`
3. Apply the changeset versioning locally: `npx changeset version`
4. Run the build `pnpm build`
5. Start a new verdaccio instance and login (see https://www.notion.so/nomicfoundation/Using-verdaccio-e15e379b59a64ef8a82fe0c1a6b93dd8?pvs=4)
6. publish `hardhat` to verdaccio:
```shell
cd packages/hardhat-core
pnpm publish --registry=http://localhost:4873/ --no-git-check
```
7. publish `@nomicfoundation/hardhat-toolbox` to verdaccio:
```shell
cd packages/hardhat-toolbox
pnpm publish --registry=http://localhost:4873/ --no-git-check
```
8. publish `@nomicfoundation/hardhat-toolbox-viem` to verdaccio:
```shell
cd packages/hardhat-toolbox-viem
pnpm publish --registry=http://localhost:4873/ --no-git-check
```
9. Install the new hardhat globally from verdaccio (is there a better approach? Also you might want to do this outside a volta controller directory):
```shell
npm install -g hardhat@2.22.1 --registry=http://localhost:4873/
```
10. Create your new sample project with the globally installed hardhat:
```shell
hardhat --version
# 2.20.2
echo "registry=http://localhost:4873" > .npmrc
hardhat init
```
11. Run the docs locally:
```shell
cd docs
pnpm install
pnpm dev
```

## Questions and decisions

- [x] Are we happy with the failing CI Test - yes a local manual run with updated versions counts as a pass
- [x] How do we approach scripting in the docs given that it is currently about deploymeent
  - Remove the scripts from the sample projects
  - Remove the scripting guide from basic, move it to advanced and show a replacement script story (e.g. "how long left in timelock")
- [x] Confirm whether we are updating the verification guide - yes
  - short intro going through ignition verification
  - follow on section for manual verification based on existing guide
- [x] Are we updating the deployment contracts to reflect the danger of msg.sender in a create2 context?
  - no - lets see if anyone raises it as an issue and reconsider

## TODO

- [x] Confirm whether we are updating the hackathon guide
- [x] Manually test the new sample projects
- [x] Investigate `assert.isDefined` issue - see comment in PR
- [x] Investigate no `ignition` in tasks issue - see comment in PR
- [x] Update verification guide in HH
- [x] Update docs to move scripts from basic to advanced
- [x] Remove scripts from sample project + update their readmes
- [x] Update sample projects with a parameters json file
- [x] Review example to work better with Ignition, specifically it is a complicated example that needs lots of Ignition Module features explained.
- [x] Run through the Tutorial section of the docs manually @ChristopherDedominici 
- [x] Run through the Guides section of the docs manually, key areas of focus: @schaable 
  - [x] Setting up a project
  - [x] Testing contracts
  - [x] Deploying your contracts
  - [x] Verifying your contracts
  - [x] Writing tasks
- [x] Run through [hardhat-boilerplate](https://github.com/NomicFoundation/hardhat-boilerplate/pull/102) @kanej 
- [x] Run through the Scripts section of Advanced in the docs manually @kanej 
- [x] Local manual run of end to end tests with bumped versions
- [x] Double check typescript peer dependency constraint as raised here: https://github.com/NomicFoundation/hardhat/pull/4955 @kanej
